db_connector/db_connector.go

```diff
@@ -0,0 +1,125 @@
+// Copyright 2017, Square, Inc.
+
+// Package db_connector is a connector to query a
+// MongoDB instance with a Query object passed in.
+package db_connector
+
+import (
+	"errors"
+	"strconv"
+
+	"github.com/square/etre/query"
+	"gopkg.in/mgo.v2"
+	"gopkg.in/mgo.v2/bson"
+)
+
+// Stores config for DB and original session. Create new
+// sessions by copying the original session:
+// https://godoc.org/labix.org/v2/mgo#Session.Copy.
+type DBConnector struct {
+	URL        string
+	Database   string
+	Collection string
+	Session    *mgo.Session
+}
+
+// NewDBConnector create an instance of DBConnector,
+// which includes storing the config needed to connect
+// to DB and creating the first session.
+func NewDBConnector(url string, database string, collection string) (*DBConnector, error) {
+	session, err := mgo.Dial(url)
+	if err != nil {
+		return nil, err
+	}
+
+	return &DBConnector{
+		URL:        url,
+		Database:   database,
+		Collection: collection,
+		Session:    session,
+	}, nil
+}
+
+// This should be called when finished with db connector instance.
+func (dbc *DBConnector) CloseSession() {
+	dbc.Session.Close()
+}
+
+// Query MongoDB instance with a Query object, which is
+// tranlsated to bson.M object for mgo driver to
+// understand.
+func (dbc *DBConnector) Query(q query.Query) (bson.M, error) {
+	var err error
+	var mgoQuery bson.M
+	var entity bson.M
+
+	mgoQuery, err = translateQuery(q)
+	if err != nil {
+		return entity, err
+	}
+
+	session := dbc.Session.Copy()
+	defer session.Close()
+
+	c := session.DB(dbc.Database).C(dbc.Collection)
+	err = c.Find(mgoQuery).One(&entity)
+	if err != nil {
+		return entity, err
+	}
+
+	return entity, nil
+}
+
+// Translates Query object to bson.M object, which will
+// be used to query mongodb instance. Part of this
+// translation includes mapping Kubernetes Selection
+// Operator
+// (https://github.com/kubernetes/apimachinery/blob/master/pkg/selection/operator.go)
+// to a MongoDB Operator
+// (https://docs.mongodb.com/manual/reference/operator/query/#query-selectors).
+// It also translates the Values in the query passed in
+// as the Kubernetes Labels Selector (KLS) parsed it into a
+// set of strings, but there are some assumptions we can
+// make based off of these rules:
+// https://github.com/kubernetes/apimachinery/blob/master/pkg/labels/selector.go#L104-L110.
```

Comment:
I think this value transformation should go in the query package (as in, do the value transformation, then store that in Query object)